### PR TITLE
RPC v0.5.0

### LINKED ISSUE #1169: Remove hack in `FullNodeClient.get_block_traces`
caused by `BLOCK_HASH` in spec instead of `BLOCK_ID`

### LINKED ISSUE #1177: FIx `TransactionTraceSchema.get_data_type` method after RPC v0.5.0 drop
Change sketchy string logic to checking `"type"` property

### LINKED ISSUE #1186: Update RPC to v0.5.0
https://github.com/starkware-libs/starknet-specs/tree/v0.5.0-rc0

Related issues:
- #1177 
- #1169 
- #1140

### PR DESCRIPTION
**Pull request by tomek0123456789**
Closes #1186
Closes #1177
Closes #1169 

## Introduced changes

- added RPC 0.5.0

##

- [x] This PR contains breaking changes

- a few, find them in migration guide

### CODE REVIEW

File: docs/migration_guide.rst [hunk 1]
```diff
@@ -5,21 +5,40 @@ Migration guide
...
-Version 0.18.3 of **starknet.py** comes with support for ...
+Version 0.18.3 of **starknet.py** comes with support for RPC 0.5.0!
 
 
 0.18.3 Targeted versions
 ------------------------
 
 - Starknet - `0.12.2 <https://community.starknet.io/t/introducing-p2p-authentication-and-mismatch-resolution-in-v0-12-2/97993>`_
-- RPC - `0.4.0 <https://github.com/starkware-libs/starknet-specs/releases/tag/v0.4.0>`_
+- RPC - `0.5.0 <https://github.com/starkware-libs/starknet-specs/releases/tag/v0.4.0>`_
```
Comment (ddoktorski) [inline docs/migration_guide.rst:15]: nit: The link is not updated

File: starknet_py/net/client_models.py [hunk 2]
```diff
@@ -279,13 +312,31 @@ class BlockStatus(Enum):
...
+    """
+    Dataclass representing a pending block on Starknet containing transaction hashes.
+    """
+
+    transactions: List[int]
+    parent_block_hash: int
+    timestamp: int
+    sequencer_address: int
+    l1_gas_price: Optional[ResourcePrice] = None
+    starknet_version: Optional[str] = None
```
Comment (ddoktorski) [inline starknet_py/net/client_models.py:339]: nit: I guess to be consistent a comment `TODO (#1179): this field should be required` should be also added here

File: starknet_py/net/client_models.py [hunk 3]
```diff
@@ -71,8 +71,28 @@ class L2toL1Message:
     """
 
     payload: List[int]
-    l1_address: int
-    l2_address: int
+    l1_address: int  # from_address in spec
+    l2_address: int  # to_address in spec
```
Comment (drknzz) [inline starknet_py/net/client_models.py:75]: If it's L2 -> L1 message, shouldn't `from_address` be `l2_address`?

File: starknet_py/net/client_models.py [hunk 4]
```diff
@@ -706,17 +761,59 @@ class SignatureOnStateDiff:
...
+
+@dataclass
+class OrderedMessage:
+    """
+    Dataclass representing a message alongside its order within the transaction.
+    """
+
+    payload: List[int]
+    l1_address: int  # from_address in spec
+    l2_address: int  # to_address in spec
```
Comment (drknzz) [inline starknet_py/net/client_models.py:815]: Same here

File: starknet_py/net/client_models.py (hunk 3 above)
Comment (tomek0123456789) [reply starknet_py/net/client_models.py:75]: Yup, thanks, good catch!

**Pull request merged**